rust/connlib/tunnel/src/io.rs

```diff
@@ -126,15 +126,12 @@ impl Io {
         self.upstream_dns_servers = create_resolvers(dns_servers);
     }
 
-    pub fn perform_dns_query(&mut self, query: DnsQuery<'static>) {
+    pub fn perform_dns_query(&mut self, query: DnsQuery<'static>) -> Result<(), DnsQueryError> {
```

Comment:
I'm guessing from the signature that this feeds into `on_dns_result` somehow. None of these functions take time, so we already know from the query packet and internal state whether it will be valid.

Reply:
Yes. Basically, what we do is we pass information back and forth between `ClientState`, which is the sans-IO component of the `Tunnel` and `Io` which is the part that causes the side-effects requested from `ClientState`.